File: scripts/safe_rename.py
import re
import requests
from datetime import date

redirect_text = '''
[[redirects]]
    from = "{}"
    to = "{}"
'''

redirect_with_date_comment_text = '''
# Added: {}{}'''

netlify_config_text = None
local_netlify_config_text = None

# Rules
def same_path(from_path, to_path, new_redirect):
    return from_path == to_path

def md_to_mdx(from_path, to_path, new_redirect):
    # old rule
    # return '.mdx' not in from_paths[i] and '.mdx' in to_paths[i]
    return from_path.endswith('.md') and to_path.endswith('.mdx')

def redirect_exists(from_path, to_path, new_redirect):
    return new_redirect in local_netlify_config_text

def from_dot_star(from_path, to_path, new_redirect):
    return from_path == '(.*)'

def is_snippets_rename(from_path, to_path, new_redirect):
    return '/snippets/' in from_path or '/snippets' in to_path

skip_rules = [
    same_path,
    md_to_mdx,
    redirect_exists,
    from_dot_star,
    is_snippets_rename
]

try:
    with open("./pr_diff", "r") as git_diff_file:
        git_diff = git_diff_file.read()

    rename_from_regex = r'rename from contents(.*).md'
    rename_to_regex = r'rename to contents(.*).md'

    from_paths = re.findall(rename_from_regex, git_diff)
    to_paths = re.findall(rename_to_regex, git_diff)

    if len(from_paths) > 0 and len(from_paths) == len(to_paths):
        new_redirects = ''

        for i in range(len(from_paths)):
            # Load existing remote redirect config
            netlify_config_text = requests.get('https://raw.githubusercontent.com/PostHog/posthog.com/master/netlify.toml').text

            # Load existing redirect file to be used to avoid duplicates
            local_netlify_config_file = open("./netlify.toml", "r")
            local_netlify_config_text = local_netlify_config_file.read()
            local_netlify_config_file.close()

            # handle index default directory files. /path/index will become /path
            from_paths[i] = re.sub("\/index$", "", from_paths[i])
            to_paths[i] = re.sub("\/index$", "", to_paths[i])
            
            new_redirect = redirect_text.format(from_paths[i], to_paths[i])

            print(f'Testing if redirects are required for: "{from_paths[i]}" to "{to_paths[i]}"')
            skip_redirect = False
            for skip_rule in skip_rules:
                if(not skip_redirect):
                    skip_redirect = skip_rule(from_paths[i], to_paths[i], new_redirect)
                    print(f'Rule: "{skip_rule.__name__}"', 'skip_redirect?', skip_redirect)
                    
            if not skip_redirect:
                print("Creating redirect: ", new_redirect)
                new_redirects += redirect_with_date_comment_text.format(date.today(), new_redirect)
            else:
                print('Not including redirect:', new_redirect)
        
        with open("./netlify.toml", "a") as netlify_config:
            netlify_config.write(new_redirects)
except Exception as e:
    print(e)

File: scripts/safe_rename.js
/* eslint-disable @typescript-eslint/no-var-requires */
const fs = require('fs')
const fetch = require('node-fetch')

let DEBUG = false

const log = (...args) => {
    if (DEBUG) {
        console.log.apply(null, args)
    }
}

/** Return date (without time) in ISO format. */
const formatDate = (date) => {
    return date.toISOString().slice(0, 10)
}

const redirectText = (from, to) => {
    return `
[[redirects]]
    from = "${from}"
    to = "${to}"
`
}

const redirectWithDateCommentText = (date, redirect) => {
    return `

# Added: ${formatDate(date)}${redirect}`
}

// Rules
const samePath = ({ fromPath, toPath }) => {
    return fromPath === toPath
}

const mdToMdx = ({ fromPath, toPath }) => {
    // # old rule
    // # return '.mdx' not in from_paths[i] and '.mdx' in to_paths[i]
    return fromPath.endsWith('.md') && toPath.endsWith('.mdx')
}

const redirectExists = ({ redirect, localConfig, remoteConfig }) => {
    return localConfig.indexOf(redirect.trim()) !== -1 || remoteConfig.indexOf(redirect.trim()) !== -1
}

const fromDotStar = ({ fromPath }) => {
    return fromPath == '(.*)'
}

const isSnippetsRename = ({ fromPath, toPath }) => {
    return fromPath.indexOf('/snippets/') !== -1 || toPath.indexOf('/snippets') !== -1
}

const skipRules = {
    samePath,
    mdToMdx,
    redirectExists,
    fromDotStar,
    isSnippetsRename,
}

// # Load existing remote redirect config
const getRemoteConfig = async () => {
    const response = await fetch('https://raw.githubusercontent.com/PostHog/posthog.com/master/netlify.toml')
    return await response.text()
}

// # Load existing redirect file to be used to avoid duplicates
const getLocalConfig = async () => {
    return await fs.promises.readFile('./netlify.toml', 'utf8')
}

const appendToLocalConfig = async (newRedirects) => {
    await fs.promises.appendFile('./netlify.toml', newRedirects)
}

const getRedirects = async ({ gitDiff, localConfig, remoteConfig, debug = false }) => {
    DEBUG = debug
    log(gitDiff)

    const renameFromRegex = new RegExp('rename from contents(.*).md', 'g')
    const renameToRegex = new RegExp('rename to contents(.*).md', 'g')

    const fromPaths = Array.from(gitDiff.matchAll(renameFromRegex), (m) => m[1])
    const toPaths = Array.from(gitDiff.matchAll(renameToRegex), (m) => m[1])

    log(fromPaths, toPaths)

    const newRedirects = []
    if (fromPaths.length > 0 && fromPaths.length === toPaths.length) {
        for (let i = 0; i < fromPaths.length; ++i) {
            // handle index default directory files. /path/index will become /path
            const fromPath = fromPaths[i].replace(/\/index$/, '', fromPaths[i])
            const toPath = toPaths[i].replace(/\/index$/, '', toPaths[i])

            const redirect = redirectText(fromPath, toPath)
            // console.log(redirect)

            log(`Testing if redirects are required for: "${fromPath}" to "${toPath}"`)

            let skipRedirect = false
            for (const [skipRuleName, skipRuleFunction] of Object.entries(skipRules)) {
                if (skipRedirect === false) {
                    skipRedirect = skipRuleFunction({ fromPath, toPath, redirect, localConfig, remoteConfig })
                    log(`Rule: "${skipRuleName}"', 'skipRedirect?', ${skipRedirect}`)
                }
            }

            if (skipRedirect === false) {
                newRedirects.push(redirectWithDateCommentText(new Date(), redirect))
            } else {
                log('Not including redirect', redirect)
            }
        }
    } else {
        log('No path changes found')
    }

    return newRedirects
}

const main = async () => {
    try {
        const gitDiff = await fs.promises.readFile('./pr_diff', 'utf8')

        const remoteConfig = await getRemoteConfig()
        const localConfig = await getLocalConfig()
        const debug = process.env.DEBUG
        const redirects = await getRedirects({ gitDiff, localConfig, remoteConfig, debug })

        if (redirects.length > 0) {
            log('Writing', redirects)
            await appendToLocalConfig(redirects.join(''))
        }
    } catch (error) {
        console.error(error)
    }
}

if (process.env.RUN_AS_SCRIPT === 'true') {
    main()
}

module.exports = {
    getRedirects,
}
